ssl/t1_lib.c

```diff
@@ -3452,3 +3450,27 @@ size_t ssl_hmac_size(const SSL_HMAC *ctx)
     return 0;
 }
 
+#ifndef OPENSSL_NO_EC
+int ssl_get_EC_curve_nid(const EVP_PKEY *pkey)
+{
+    int ret = NID_undef;
+
+    if (EVP_PKEY_id(pkey) != EVP_PKEY_KEYMGMT) {
+        if (EVP_PKEY_base_id(pkey) == EVP_PKEY_EC) {
+            EC_KEY *ec = EVP_PKEY_get0_EC_KEY(pkey);
+
+            ret = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec));
+        }
+    } else {
+        char name[80] = { 0 };
+        size_t name_len;
+
+        if (EVP_PKEY_get_utf8_string_param((EVP_PKEY *)pkey,
+                                           OSSL_PKEY_PARAM_GROUP_NAME,
+                                           name, sizeof(name), &name_len))
+            ret = EC_curve_name2nid(name);
```

Comment:
I think this is the wrong approach. Rather than indexing on nid we could index on the TLS_GROUP_INFO entry in the ctx->group_list array. We should scan that list looking for a "realname" that matches the curve name we've just got from the provider.

Reply:
> I think this is the wrong approach. Rather than indexing on nid we could index on the TLS_GROUP_INFO entry in the ctx->group_list array. We should scan that list looking for a "realname" that matches the curve name we've just got from the provider.

Do we have all those details in place?  If not, that must happen first.  All in all, though, I completely agree with you.

Reply:
> Do we have all those details in place? If not, that must happen first. All in all, though, I completely agree with you.

Yes. This is the basis of the capabilities work I did. We ask the providers what groups they support using the "TLS-GROUP" capability and populate `ctx->group_list` with everything we found.

Reply:
I've kept this to be for NIDs for now...  The rest is a much bigger refactor that I think needs to be done separately.